File: src/test/java/org/hisp/dhis/rules/RuleEngineValueTypesTests.java
package org.hisp.dhis.rules;

import org.hisp.dhis.rules.models.*;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.*;

import static com.google.common.truth.Truth.assertThat;

@RunWith( JUnit4.class )
public class RuleEngineValueTypesTests
{
        @Test
        public void booleanVariableWithoutValueMustFallbackToDefaultBooleanValue()
            throws Exception
        {
                RuleAction ruleAction = RuleActionDisplayKeyValuePair
                    .createForFeedback( "test_action_content", "#{test_variable}" );
                Rule rule = Rule.create( null, null, "true", Arrays.asList( ruleAction ), "");
                RuleVariable ruleVariable = RuleVariableCurrentEvent
                    .create( "test_variable", "test_data_element", RuleValueType.BOOLEAN );

                RuleEngine ruleEngine = getRuleEngine( rule, Arrays.asList( ruleVariable ) );

                RuleEvent ruleEvent = RuleEvent.create( "test_event", "test_program_stage",
                    RuleEvent.Status.ACTIVE, new Date(), new Date(), "",null, new ArrayList<RuleDataValue>(), "");
                List<RuleEffect> ruleEffects = ruleEngine.evaluate( ruleEvent ).call();

                assertThat( ruleEffects.size() ).isEqualTo( 1 );
                assertThat( ruleEffects.get( 0 ).data() ).isEqualTo( "false" );
                assertThat( ruleEffects.get( 0 ).ruleAction() ).isEqualTo( ruleAction );
        }

        @Test
        public void numericVariableWithoutValueMustFallbackToDefaultNumericValue()
            throws Exception
        {
                RuleAction ruleAction = RuleActionDisplayKeyValuePair
                    .createForFeedback( "test_action_content", "#{test_variable}" );
                Rule rule = Rule.create( null, null, "true", Arrays.asList( ruleAction ), "");
                RuleVariable ruleVariable = RuleVariableCurrentEvent
                    .create( "test_variable", "test_data_element", RuleValueType.NUMERIC );

                RuleEngine ruleEngine = getRuleEngine( rule, Arrays.asList( ruleVariable ) );

                RuleEvent ruleEvent = RuleEvent.create( "test_event", "test_program_stage",
                    RuleEvent.Status.ACTIVE, new Date(), new Date(), "", null,new ArrayList<RuleDataValue>(), "");
                List<RuleEffect> ruleEffects = ruleEngine.evaluate( ruleEvent ).call();

                assertThat( ruleEffects.size() ).isEqualTo( 1 );
                assertThat( ruleEffects.get( 0 ).data() ).isEqualTo( "0.0" );
                assertThat( ruleEffects.get( 0 ).ruleAction() ).isEqualTo( ruleAction );
        }

        @Test
        public void textVariableWithoutValueMustFallbackToDefaultTextValue()
            throws Exception
        {
                RuleAction ruleAction = RuleActionDisplayKeyValuePair
                    .createForFeedback( "test_action_content", "#{test_variable}" );
                Rule rule = Rule.create( null, null, "true", Arrays.asList( ruleAction ), "");
                RuleVariable ruleVariable = RuleVariableCurrentEvent
                    .create( "test_variable", "test_data_element", RuleValueType.TEXT );

                RuleEngine ruleEngine = getRuleEngine( rule, Arrays.asList( ruleVariable ) );

                RuleEvent ruleEvent = RuleEvent.create( "test_event", "test_program_stage",
                    RuleEvent.Status.ACTIVE, new Date(), new Date(), "",null, new ArrayList<RuleDataValue>(), "");
                List<RuleEffect> ruleEffects = ruleEngine.evaluate( ruleEvent ).call();

                assertThat( ruleEffects.size() ).isEqualTo( 1 );
                assertThat( ruleEffects.get( 0 ).data() ).isEqualTo( "" );
                assertThat( ruleEffects.get( 0 ).ruleAction() ).isEqualTo( ruleAction );
        }

        private RuleEngine getRuleEngine( Rule rule, List<RuleVariable> ruleVariables )
        {
                return RuleEngineContext
                        .builder( new ExpressionEvaluator() )
                        .rules( Arrays.asList( rule ) )
                        .ruleVariables( ruleVariables )
                        .calculatedValueMap( new HashMap<>( ) )
                        .supplementaryData( new HashMap<>() )
                        .constantsValue( new HashMap<>() )
                        .build().toEngineBuilder().triggerEnvironment( TriggerEnvironment.SERVER )
                        .build();
        }
}

File: src/jvmTest/kotlin/org/hisp/dhis/rules/RuleEngineValueTypesTests.kt
package org.hisp.dhis.rules

import org.assertj.core.api.Assertions.assertThat
import org.hisp.dhis.rules.models.*
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.util.*

@RunWith(JUnit4::class)
class RuleEngineValueTypesTests {

    @Test
    @Throws(Exception::class)
    fun booleanVariableWithoutValueMustFallbackToDefaultBooleanValue() {
        val ruleAction = RuleActionDisplayKeyValuePair
                .createForFeedback("test_action_content", "#{test_variable}")
        val rule = Rule.create(null, null, "true", listOf(ruleAction), "")
        val ruleVariable = RuleVariableCurrentEvent
                .create("test_variable", "test_data_element", RuleValueType.BOOLEAN)

        val ruleEngine = getRuleEngine(rule, listOf(ruleVariable))

        val ruleEvent = RuleEvent.create("test_event", "test_program_stage",
                RuleEvent.Status.ACTIVE, Date(), Date(), "", null, ArrayList(), "")
        val ruleEffects = ruleEngine.evaluate(ruleEvent).call()

        assertThat(ruleEffects.size).isEqualTo(1)
        assertThat(ruleEffects[0].data).isEqualTo("false")
        assertThat(ruleEffects[0].ruleAction).isEqualTo(ruleAction)
    }

    @Test
    @Throws(Exception::class)
    fun numericVariableWithoutValueMustFallbackToDefaultNumericValue() {
        val ruleAction = RuleActionDisplayKeyValuePair
                .createForFeedback("test_action_content", "#{test_variable}")
        val rule = Rule.create(null, null, "true", listOf(ruleAction), "")
        val ruleVariable = RuleVariableCurrentEvent
                .create("test_variable", "test_data_element", RuleValueType.NUMERIC)

        val ruleEngine = getRuleEngine(rule, listOf(ruleVariable))

        val ruleEvent = RuleEvent.create("test_event", "test_program_stage",
                RuleEvent.Status.ACTIVE, Date(), Date(), "", null, ArrayList(), "")
        val ruleEffects = ruleEngine.evaluate(ruleEvent).call()

        assertThat(ruleEffects.size).isEqualTo(1)
        assertThat(ruleEffects[0].data).isEqualTo("0.0")
        assertThat(ruleEffects[0].ruleAction).isEqualTo(ruleAction)
    }

    @Test
    @Throws(Exception::class)
    fun textVariableWithoutValueMustFallbackToDefaultTextValue() {
        val ruleAction = RuleActionDisplayKeyValuePair
                .createForFeedback("test_action_content", "#{test_variable}")
        val rule = Rule.create(null, null, "true", listOf(ruleAction), "")
        val ruleVariable = RuleVariableCurrentEvent
                .create("test_variable", "test_data_element", RuleValueType.TEXT)

        val ruleEngine = getRuleEngine(rule, listOf(ruleVariable))

        val ruleEvent = RuleEvent.create("test_event", "test_program_stage",
                RuleEvent.Status.ACTIVE, Date(), Date(), "", null, ArrayList(), "")
        val ruleEffects = ruleEngine.evaluate(ruleEvent).call()

        assertThat(ruleEffects.size).isEqualTo(1)
        assertThat(ruleEffects[0].data).isEqualTo("")
        assertThat(ruleEffects[0].ruleAction).isEqualTo(ruleAction)
    }

    private fun getRuleEngine(rule: Rule, ruleVariables: List<RuleVariable>): RuleEngine {
        return RuleEngineContext
                .builder(ExpressionEvaluator())
                .rules(listOf(rule))
                .ruleVariables(ruleVariables)
                .calculatedValueMap(hashMapOf())
                .supplementaryData(hashMapOf())
                .constantsValue(hashMapOf())
                .build().toEngineBuilder().triggerEnvironment(TriggerEnvironment.SERVER)
                .build()
    }
}
